kit/tracing/tracing.go

```diff
@@ -44,21 +43,60 @@ func ExtractFromHTTPRequest(req *http.Request, handlerName string) (opentracing.
 	return span, req
 }
 
-// StartSpanFromContext is an easier-to-use opentracing.StartSpanFromContext.
-// Uses the calling function as the operation name, and logs the file:line.
-// TODO unused for now; consider removing this because performance isn't awesome.
+// StartSpanFromContext is an improved opentracing.StartSpanFromContext.
+// Uses the calling function as the operation name, and logs the filename and line number.
+//
+// Passing nil context triggers new context and root span construction.
+// Context without parent span reference triggers root span construction.
+// This function never returns nil values.
+//
+// Performance
+//
+// This function incurs a small performance penalty, roughly 1000 ns/op, 376 B/op, 6 allocs/op.
+// Jaeger timestamp and duration precision is only µs, so this is pretty negligible.
+//
+// Alternatives
+//
+// If this performance penalty is too much, try these, which are also demonstrated in benchmark tests:
+//  // Create a root span
+//  span := opentracing.StartSpan("operation name")
+//
+//  // Create a child span
+//  span := opentracing.StartSpan("operation name", opentracing.ChildOf(sc))
+//
+//  // Sugar to create a child span
+//  span, ctx := opentracing.StartSpanFromContext(ctx, "operation name")
 func StartSpanFromContext(ctx context.Context) (opentracing.Span, context.Context) {
-	pc, _, _, ok := runtime.Caller(1)
-	if !ok {
-		span, ctx := opentracing.StartSpanFromContext(ctx, "unknown")
-		span.LogFields(log.Error(errors.New("failed to get calling frame")))
-		return span, ctx
+	if ctx == nil {
+		// Guard against nil context.
+		ctx = context.Background()
```

Comment:
I would argue that this is a suitable place to panic. I would rather be informed that I made a mistake in calling StartSpanFromContext with a nil context, so that I have to decide whether to instantiate my own context.Background, or so that I can figure out why I had a nil context when I didn't expect one.

Reply:
I can get behind that.